File: chapitre-10/exemple-nice.c
// ------------------------------------------------------------------
// exemple-nice.c
// Fichier d'exemple du livre "Developpement Systeme sous Linux"
// (C) 2000-2019 - Christophe BLAESS <christophe@blaess.fr>
// https://www.blaess.fr/christophe/
// ------------------------------------------------------------------

#define _GNU_SOURCE
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <sys/wait.h>
#include <unistd.h>

volatile long counter = 0;
static int    nice_value;

void signal_handler(int numero)
{
	if (counter != 0) {
		fprintf(stdout, "Child %ld (nice %2d) Counter = %10ld\n",
		                 (long) getpid(), nice_value, counter);
		exit(EXIT_SUCCESS);
	}
}

#define NB_CHILD    5

int main (void)
{
	pid_t   pid;
	int     child;

	/* Creation d'un nouveau groupe de processus */
	setpgid(0, 0);

	signal(SIGUSR1, signal_handler);
	for (child = 0; child < NB_CHILD; child ++) {
		if ((pid = fork()) < 0) {
			perror("fork");
			exit(EXIT_FAILURE);
		}
		if (pid != 0)
			continue;
		nice_value = child * (20 / (NB_CHILD - 1));
		if (nice(nice_value) < 0) {
			perror("nice");
			exit(EXIT_FAILURE);
		}
		/* attente signal de demarrage */
		pause();
		/* comptage */
		while (1)
			counter ++;
	}
	/* processus parent */
	signal(SIGUSR1, SIG_IGN);
	sleep(1);
	kill(-getpgid(0), SIGUSR1);
	sleep(5);
	kill(-getpgid(0), SIGUSR1);
	while (wait(NULL) > 0)
		;
	exit(EXIT_SUCCESS);
}

File: chapitre-10/exemple-nice.py
#! /usr/bin/python
# ------------------------------------------------------------------
# exemple-nice.py
# Fichier d'exemple du livre "Developpement Systeme sous Linux"
# (C) 2015-2019 - Christophe BLAESS <christophe@blaess.fr>
# https://www.blaess.fr/christophe/
# ------------------------------------------------------------------
from __future__ import print_function
import os
import signal
import sys
import time

count = 0;
child = 0
nice = 0

def handler_usr1(signum, stack):
	print("Child {}, nice={:2}, count={}"
	      .format(child, nice, count))


os.setpgid(0, 0)

for child in range(0, 5):
	pid = os.fork()
	if (pid != 0):
		continue
	# Child branch
	signal.signal(signal.SIGUSR1, handler_usr1)
	nice = child * 4
	os.nice(nice)
	signal.pause()
	while True:
		count = count + 1

# Parent branch
signal.signal(signal.SIGUSR1, signal.SIG_IGN)
time.sleep(1)
os.kill(-os.getpgid(0), signal.SIGUSR1)
time.sleep(5)
os.kill(-os.getpgid(0), signal.SIGUSR1)
time.sleep(1)
os.kill(-os.getpgid(0), signal.SIGKILL)

